File: ElevatorSimulation.java
import java.util.*;

public class ElevatorSimulation {
    public static final int numFloors = 150;
    public static final int numElev = 8;
    public static final int simTime = 3600;
    public static final int capacity = 20;
    public static final int openTime = 10;

    public long totalWaitTime = 0;
    public long totalTravelTime = 0;
    public long totalDistance = 0;
    public int numPassengers = 0;

    private List<Elevator> elevators;
    private List<Floor> floors;

    public ElevatorSimulation() {
        elevators = new ArrayList<>();
        floors = new ArrayList<>();

        for (int i = 0; i < numElev; i++) {
            elevators.add(new Elevator(i, this));
        }
        for (int i = 0; i < numFloors; i++) {
            floors.add(new Floor(i));
        }
    }

    public Floor getFloor(int index) {
        if (index >= 0 && index < floors.size()) {
            return floors.get(index);
        }
        return null;
    }

    private void runSim() {
        Random random = new Random();
        
        for (int time = 0; time < simTime; time++) {
            // People entering building
            int numUp = random.nextInt(5);
            for (int i = 0; i < numUp; i++) {
                int dest = random.nextInt(numFloors - 1) + 1;
                Passenger p = new Passenger(0, dest, time);
                floors.get(0).enqueue(p);
            }

            // People leaving building
            int numDown = random.nextInt(5);
            for (int i = 0; i < numDown; i++) {
                int start = random.nextInt(numFloors - 1) + 1;
                Passenger p = new Passenger(start, 0, time);
                floors.get(start).enqueue(p);
            }
            
            processRequests();
            updateElevators(time);
        }
        printData();
    }

    private void processRequests() {
        for (Floor f : floors) {
            if (f.hasWaiting()) {
                Passenger p = f.peek();
                // Determine direction based on the first person in line
                boolean goingUp = p.getEnd() > p.getStart();
                
                Elevator bestElevator = findBestElevator(f.getNum(), goingUp);
                
                if (bestElevator != null && bestElevator.canAcceptRequest()) {
                    bestElevator.addRequest(f.getNum(), goingUp);
                }
            }
        }
    }

    private Elevator findBestElevator(int floor, boolean goingUp) {
        Elevator best = null;
        int bestScore = Integer.MAX_VALUE;

        for (Elevator e : elevators) {
            int score = e.calculateScore(floor, goingUp);
            if (score < bestScore) {
                bestScore = score;
                best = e;
            }
        }
        return best;
    }

    public void updateElevators(int currTime) {
        for (Elevator e : elevators) {
            e.update(currTime);
        }
    }

    public void recordPassenger(Passenger p) {
        totalWaitTime += p.getWait();
        totalTravelTime += p.getTotalTime();
        numPassengers++;
    }
    
    public void recordDistance() {
        totalDistance++;
    }

    private void printData() {
        System.out.println("--- Simulation Results ---");
        System.out.println("Total passengers delivered: " + numPassengers);
        
        // Debug info
        int totalWaiting = 0;
        for (Floor f : floors) {
            totalWaiting += f.getWaitingCount();
        }
        System.out.println("Passengers still waiting: " + totalWaiting);
        
        // Elevator status
        System.out.println("\nElevator Status:");
        for (Elevator e : elevators) {
            e.printStatus();
        }
        
        if (numPassengers > 0) {
            System.out.println("\nAvg wait time: " + (totalWaitTime / numPassengers) + " sec");
            System.out.println("Avg total trip time: " + (totalTravelTime / numPassengers) + " sec");
        }
        System.out.println("\nAvg distance per elevator: " + (totalDistance / numElev) + " floors");
    }

    public static void main(String[] args) {
        new ElevatorSimulation().runSim();
    }
}

class Passenger {
    private int start;
    private int end;
    private int requestTime;
    private int enterTime;
    private int exitTime;

    public Passenger(int start, int end, int currTime) {
        this.start = start;
        this.end = end;
        this.requestTime = currTime;
    }

    public int getStart() { return start; }
    public int getEnd() { return end; }

    public void enter(int time) { enterTime = time; }
    public void exit(int time) { exitTime = time; }

    public int getWait() { return enterTime - requestTime; }
    public int getTotalTime() { return exitTime - requestTime; }
}

class Elevator {
    private int num;
    private int currFloor;
    private int doorTimer;
    private List<Passenger> passengers;
    private State state;
    private ElevatorSimulation sim;
    
    // Using lists to track stops
    private List<Integer> upStops;
    private List<Integer> downStops;

    public enum State {
        IDLE, UP, DOWN, OPEN
    }

    public Elevator(int id, ElevatorSimulation sim) {
        this.num = id;
        this.sim = sim;
        this.state = State.IDLE;
        this.currFloor = 0;
        this.passengers = new ArrayList<>();
        this.upStops = new ArrayList<>();
        this.downStops = new ArrayList<>();
    }

    public boolean canBoard() {
        return passengers.size() < ElevatorSimulation.capacity;
    }

    public boolean canAcceptRequest() {
        // Limit queue to prevent one elevator from hogging all requests
        return state == State.IDLE || (upStops.size() + downStops.size() < 50);
    }

    public void addRequest(int floor, boolean goingUp) {
        // Add the floor based on what direction need to go
        
        if (floor > currFloor) {
            // Floor is above, add to upStops
            if (!upStops.contains(floor)) {
                upStops.add(floor);
            }
        } else if (floor < currFloor) {
            // Floor is below , add to downStops
            if (!downStops.contains(floor)) {
                downStops.add(floor);
            }
        } else {
            // At this floor, open doors
            if (state == State.IDLE) {
                state = State.OPEN;
                doorTimer = ElevatorSimulation.openTime;
            }
        }
        updateState();
    }

    private void updateState() {
        if (state == State.IDLE) {
            if (!upStops.isEmpty()) {
                state = State.UP;
            } else if (!downStops.isEmpty()) {
                state = State.DOWN;
            }
        }
    }

    // Heuristic function to choose best elevator
    public int calculateScore(int targetFloor, boolean goingUp) {
        if (state == State.IDLE) {
            return Math.abs(currFloor - targetFloor);
        } 
        // Going up and the request is above
        else if (state == State.UP && goingUp && targetFloor >= currFloor) {
            return targetFloor - currFloor;
        } 
        // Going down and the request is below
        else if (state == State.DOWN && !goingUp && targetFloor <= currFloor) {
            return currFloor - targetFloor;
        } 
        else {
            // Penalty if wrong direction
            return 1000 + Math.abs(currFloor - targetFloor);
        }
    }

    public void update(int time) {
        switch (state) {
            case UP:
                moveUp(time);
                break;
            case DOWN:
                moveDown(time);
                break;
            case OPEN:
                handleDoors(time);
                break;
            case IDLE:
                updateState(); // Check if new tasks
                break;
        }
    }

    private void moveUp(int time) {
        if (currFloor >= ElevatorSimulation.numFloors - 1) {
            determineNextState();
            return;
        }
        
        currFloor++;
        sim.recordDistance();
        
        // Check if need to stop at this floor
        boolean shouldStop = false;
        
        for (Passenger p : passengers) {
            if (p.getEnd() == currFloor) {
                shouldStop = true;
                break;
            }
        }
        
        // Check if this floor has pickup request
        if (upStops.contains(currFloor)) {
            shouldStop = true;
            upStops.remove((Integer) currFloor);
        }
        
        if (shouldStop) {
            state = State.OPEN;
            doorTimer = ElevatorSimulation.openTime;
        } else if (upStops.isEmpty() && !hasPassengersGoingUp()) {
            // No more up destinations
            determineNextState();
        }
    }

    private void moveDown(int time) {
        if (currFloor <= 0) {
            determineNextState();
            return;
        }
        
        currFloor--;
        sim.recordDistance();
        
        // Check if need to stop at this floor
        boolean shouldStop = false;
        
        // Check if passenger needs to get off
        for (Passenger p : passengers) {
            if (p.getEnd() == currFloor) {
                shouldStop = true;
                break;
            }
        }
        
        // Check if this floor has pickup request
        if (downStops.contains(currFloor)) {
            shouldStop = true;
            downStops.remove((Integer) currFloor);
        }
        
        if (shouldStop) {
            state = State.OPEN;
            doorTimer = ElevatorSimulation.openTime;
        } else if (downStops.isEmpty() && !hasPassengersGoingDown()) {
            // No more down destinations
            determineNextState();
        }
    }
    
    private boolean hasPassengersGoingUp() {
        for (Passenger p : passengers) {
            if (p.getEnd() > currFloor) {
                return true;
            }
        }
        return false;
    }
    
    private boolean hasPassengersGoingDown() {
        for (Passenger p : passengers) {
            if (p.getEnd() < currFloor) {
                return true;
            }
        }
        return false;
    }

    private void handleDoors(int time) {
        // 1. Let people off
        unload(time);

        // 2. Let people on
        Floor f = sim.getFloor(currFloor);
        if (f != null) {
            // Determine next direction based on remaining stops
            boolean willGoUp = !upStops.isEmpty();
            boolean willGoDown = !downStops.isEmpty();
            
            while (canBoard() && f.hasWaiting()) {
                Passenger p = f.peek();
                boolean passengerGoingUp = p.getEnd() > p.getStart();
                
                // Board if: going our direction, or we have no direction yet
                if ((willGoUp && passengerGoingUp) || 
                    (willGoDown && !passengerGoingUp) ||
                    (!willGoUp && !willGoDown)) {
                    p = f.dequeue();
                    p.enter(time);
                    passengers.add(p);
                    
                    if (passengerGoingUp) {
                        if (!upStops.contains(p.getEnd())) {
                            upStops.add(p.getEnd());
                        }
                    } else {
                        if (!downStops.contains(p.getEnd())) {
                            downStops.add(p.getEnd());
                        }
                    }
                    
                    // Update direction flags as new passengers board
                    willGoUp = !upStops.isEmpty();
                    willGoDown = !downStops.isEmpty();
                } else {
                    break; // Next passenger wants to go the other way
                }
            }
        }

        // 3. Close doors check
        doorTimer--;
        if (doorTimer <= 0) {
            determineNextState();
        }
    }

    private void unload(int time) {
        Iterator<Passenger> it = passengers.iterator();
        while (it.hasNext()) {
            Passenger p = it.next();
            if (p.getEnd() == currFloor) {
                p.exit(time);
                sim.recordPassenger(p);
                it.remove();
            }
        }
    }

    private void determineNextState() {
        // Clear any invalid stops
        upStops.removeIf(floor -> floor < 0 || floor >= ElevatorSimulation.numFloors);
        downStops.removeIf(floor -> floor < 0 || floor >= ElevatorSimulation.numFloors);
        
        if (!upStops.isEmpty()) {
            state = State.UP;
        } else if (!downStops.isEmpty()) {
            state = State.DOWN;
        } else {
            state = State.IDLE;
        }
    }
    
    public void printStatus() {
        System.out.println("Elevator " + num + ": Floor " + currFloor + ", State: " + state + 
                         ", Passengers: " + passengers.size());
        System.out.println("  UpStops: " + upStops);
        System.out.println("  DownStops: " + downStops);
    }
}

class Floor {
    private int num;
    private Queue<Passenger> waiting;

    public Floor(int num) {
        this.num = num;
        waiting = new LinkedList<>();
    }

    public int getNum() { return num; }
    public boolean hasWaiting() { return !waiting.isEmpty(); }
    public int getWaitingCount() { return waiting.size(); }
    public void enqueue(Passenger p) { waiting.add(p); }
    public Passenger dequeue() { return waiting.poll(); }
    public Passenger peek() { return waiting.peek(); }
}

File: ElevatorSimulation.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>
#include <limits.h>
#include <float.h>

#define NUM_FLOORS 150
#define NUM_ELEV 8
#define SIM_TIME 3600
#define CAPACITY 20
#define OPEN_TIME 10

// Passenger structure, does not depend on others
typedef struct Passenger {
    int start;
    int end;
    int requestTime;
    int enterTime;
    int exitTime;
} Passenger;

// Forward declarations for others
typedef struct Elevator Elevator;
typedef struct Floor Floor;
typedef struct ElevatorSimulation ElevatorSimulation;

Passenger* newPassenger(int start, int end, int currTime) {
    Passenger* p = (Passenger*)malloc(sizeof(Passenger));
    p->start = start;
    p->end = end;
    p->requestTime = currTime;
    p->enterTime = 0;
    p->exitTime = 0;
    return p;
}

// Queue for passengers
typedef struct QueueItem {
    Passenger* passenger;
    struct QueueItem* next;
} QueueItem;

typedef struct Queue {
    QueueItem* front;
    QueueItem* rear;
    int count;
} Queue;

Queue* newQueue() {
    Queue* q = (Queue*)malloc(sizeof(Queue));
    q->front = NULL;
    q->rear = NULL;
    q->count = 0;
    return q;
}

void enqueue(Queue* q, Passenger* p) {
    QueueItem* node = (QueueItem*)malloc(sizeof(QueueItem));
    node->passenger = p;
    node->next = NULL;
    
    if (q->rear == NULL) {
        q->front = q->rear = node;
    } else {
        q->rear->next = node;
        q->rear = node;
    }
    q->count++;
}

Passenger* dequeue(Queue* q) {
    if (q->front == NULL) return NULL;
    
    QueueItem* temp = q->front;
    Passenger* p = temp->passenger;
    q->front = q->front->next;
    
    if (q->front == NULL) {
        q->rear = NULL;
    }
    
    free(temp);
    q->count--;
    return p;
}

Passenger* peekQueue(Queue* q) {
    if (q->front == NULL) return NULL;
    return q->front->passenger;
}

bool isQueueEmpty(Queue* q) {
    return q->front == NULL;
}

// Array for stops
typedef struct array {
    int* data;
    int size;
    int capacity;
} array;

array* newArray() {
    array* arr = (array*)malloc(sizeof(array));
    arr->capacity = 10;
    arr->size = 0;
    arr->data = (int*)malloc(arr->capacity * sizeof(int));
    return arr;
}

void addToArray(array* arr, int value) {
    // Check if already exists
    for (int i = 0; i < arr->size; i++) {
        if (arr->data[i] == value) return;
    }
    
    if (arr->size >= arr->capacity) {
        arr->capacity *= 2;
        arr->data = (int*)realloc(arr->data, arr->capacity * sizeof(int));
    }
    
    arr->data[arr->size++] = value;
    
    // Keep sorted
    for (int i = arr->size - 1; i > 0; i--) {
        if (arr->data[i] < arr->data[i-1]) {
            int temp = arr->data[i];
            arr->data[i] = arr->data[i-1];
            arr->data[i-1] = temp;
        } else {
            break;
        }
    }
}

bool containsInt(array* arr, int value) {
    for (int i = 0; i < arr->size; i++) {
        if (arr->data[i] == value) return true;
    }
    return false;
}

void removeInt(array* arr, int value) {
    for (int i = 0; i < arr->size; i++) {
        if (arr->data[i] == value) {
            for (int j = i; j < arr->size - 1; j++) {
                arr->data[j] = arr->data[j + 1];
            }
            arr->size--;
            return;
        }
    }
}

void freeArray(array* arr) {
    free(arr->data);
    free(arr);
}

struct Floor {
    int num;
    Queue* waiting;
};

Floor* newFloor(int num) {
    Floor* f = (Floor*)malloc(sizeof(Floor));
    f->num = num;
    f->waiting = newQueue();
    return f;
}

typedef enum {
    IDLE, UP, DOWN, OPEN
} ElevatorState;

// Passenger list for elevator
typedef struct PassengerNode {
    Passenger* passenger;
    struct PassengerNode* next;
} PassengerNode;

typedef struct PassengerList {
    PassengerNode* head;
    int count;
} PassengerList;

PassengerList* createPassengerList() {
    PassengerList* list = (PassengerList*)malloc(sizeof(PassengerList));
    list->head = NULL;
    list->count = 0;
    return list;
}

void addPassenger(PassengerList* list, Passenger* p) {
    PassengerNode* node = (PassengerNode*)malloc(sizeof(PassengerNode));
    node->passenger = p;
    node->next = list->head;
    list->head = node;
    list->count++;
}

void removePassengerFromList(PassengerList* list, Passenger* p) {
    PassengerNode* curr = list->head;
    PassengerNode* prev = NULL;
    
    while (curr != NULL) {
        if (curr->passenger == p) {
            if (prev == NULL) {
                list->head = curr->next;
            } else {
                prev->next = curr->next;
            }
            free(curr);
            list->count--;
            return;
        }
        prev = curr;
        curr = curr->next;
    }
}

struct Elevator {
    int num;
    int currFloor;
    int doorTimer;
    PassengerList* passengers;
    ElevatorState state;
    array* upStops;
    array* downStops;
    int lastPickupTime;
    ElevatorSimulation* sim;
};

struct ElevatorSimulation {
    long totalWaitTime;
    long totalTravelTime;
    long totalDistance;
    int numPassengers;
    Elevator* elevators[NUM_ELEV];
    Floor* floors[NUM_FLOORS];
};

ElevatorSimulation* newSim() {
    ElevatorSimulation* sim = (ElevatorSimulation*)malloc(sizeof(ElevatorSimulation));
    sim->totalWaitTime = 0;
    sim->totalTravelTime = 0;
    sim->totalDistance = 0;
    sim->numPassengers = 0;
    
    for (int i = 0; i < NUM_FLOORS; i++) {
        sim->floors[i] = newFloor(i);
    }
    
    return sim;
}

// Declarations for Elevator functions
void updateElevState(Elevator* e);
double calculateScore(Elevator* e, int targetFloor, bool goingUp, int currentTime);
void addRequest(Elevator* e, int floor, bool goingUp);
void updateElevator(Elevator* e, int time);
void moveUp(Elevator* e, int time);
void moveDown(Elevator* e, int time);
void handleDoors(Elevator* e, int time);
int unload(Elevator* e, int time);
void getNext(Elevator* e);
bool hasUp(Elevator* e);
bool hasDown(Elevator* e);
bool isTarget(Elevator* e, int floor);
int estimateDist(Elevator* e);

Elevator* newElevator(int id, ElevatorSimulation* sim) {
    Elevator* e = (Elevator*)malloc(sizeof(Elevator));
    e->num = id;
    e->sim = sim;
    e->state = IDLE;
    e->currFloor = 0;
    e->doorTimer = 0;
    e->passengers = createPassengerList();
    e->upStops = newArray();
    e->downStops = newArray();
    e->lastPickupTime = 0;
    return e;
}

bool canBoard(Elevator* e) {
    return e->passengers->count < CAPACITY;
}

void updateElevState(Elevator* e) {
    if (e->state == IDLE) {
        if (e->upStops->size > 0) {
            e->state = UP;
        } else if (e->downStops->size > 0) {
            e->state = DOWN;
        }
    }
}

void addRequest(Elevator* e, int floor, bool goingUp) {
    if (floor > e->currFloor) {
        addToArray(e->upStops, floor);
    } else if (floor < e->currFloor) {
        addToArray(e->downStops, floor);
    } else {
        if (e->state == IDLE) {
            e->state = OPEN;
            e->doorTimer = OPEN_TIME;
        }
    }
    updateElevState(e);
}

int estimateDist(Elevator* e) {
    int dist = 0;
    if (e->upStops->size > 0) {
        dist = abs(e->currFloor - e->upStops->data[e->upStops->size - 1]);
    }
    if (e->downStops->size > 0) {
        int downDist = abs(e->currFloor - e->downStops->data[0]);
        if (downDist > dist) dist = downDist;
    }
    return dist;
}

double calculateScore(Elevator* e, int targetFloor, bool goingUp, int currentTime) {
    int totalStops = e->upStops->size + e->downStops->size;
    
    if (e->passengers->count >= CAPACITY) {
        return DBL_MAX;
    }
    
    if (e->state == IDLE) {
        if (totalStops >= 8) return DBL_MAX;
        return abs(e->currFloor - targetFloor);
    }
    
    if (totalStops >= 5) {
        return DBL_MAX;
    }
    
    if (e->state == UP && goingUp && targetFloor >= e->currFloor) {
        return (targetFloor - e->currFloor) + (e->passengers->count * 2) + (totalStops * 10);
    }
    
    if (e->state == DOWN && !goingUp && targetFloor <= e->currFloor) {
        return (e->currFloor - targetFloor) + (e->passengers->count * 2) + (totalStops * 10);
    }
    
    if (e->state == UP && goingUp && targetFloor < e->currFloor) {
        int distToFinish = estimateDist(e);
        return distToFinish + (e->currFloor - targetFloor) + (e->passengers->count * 5);
    }
    
    if (e->state == DOWN && !goingUp && targetFloor > e->currFloor) {
        int distToFinish = estimateDist(e);
        return distToFinish + (targetFloor - e->currFloor) + (e->passengers->count * 5);
    }
    
    return DBL_MAX;
}

bool hasUp(Elevator* e) {
    PassengerNode* curr = e->passengers->head;
    while (curr != NULL) {
        if (curr->passenger->end > e->currFloor) {
            return true;
        }
        curr = curr->next;
    }
    return false;
}

bool hasDown(Elevator* e) {
    PassengerNode* curr = e->passengers->head;
    while (curr != NULL) {
        if (curr->passenger->end < e->currFloor) {
            return true;
        }
        curr = curr->next;
    }
    return false;
}

bool isTarget(Elevator* e, int floor) {
    PassengerNode* curr = e->passengers->head;
    while (curr != NULL) {
        if (curr->passenger->end == floor) {
            return true;
        }
        curr = curr->next;
    }
    return false;
}

void recordPassenger(ElevatorSimulation* sim, Passenger* p) {
    sim->totalWaitTime += (p->enterTime - p->requestTime);
    sim->totalTravelTime += (p->exitTime - p->requestTime);
    sim->numPassengers++;
}

void moveUp(Elevator* e, int time) {
    if (e->currFloor >= NUM_FLOORS - 1) {
        getNext(e);
        return;
    }
    
    e->currFloor++;
    e->sim->totalDistance++;
    
    bool shouldStop = false;
    
    PassengerNode* curr = e->passengers->head;
    while (curr != NULL) {
        if (curr->passenger->end == e->currFloor) {
            shouldStop = true;
            break;
        }
        curr = curr->next;
    }
    
    if (containsInt(e->upStops, e->currFloor)) {
        shouldStop = true;
        removeInt(e->upStops, e->currFloor);
    }
    
    if (shouldStop) {
        e->state = OPEN;
        e->doorTimer = OPEN_TIME;
    } else if (e->upStops->size == 0 && !hasUp(e)) {
        getNext(e);
    }
}

void moveDown(Elevator* e, int time) {
    if (e->currFloor <= 0) {
        getNext(e);
        return;
    }
    
    e->currFloor--;
    e->sim->totalDistance++;
    
    bool shouldStop = false;
    
    PassengerNode* curr = e->passengers->head;
    while (curr != NULL) {
        if (curr->passenger->end == e->currFloor) {
            shouldStop = true;
            break;
        }
        curr = curr->next;
    }
    
    if (containsInt(e->downStops, e->currFloor)) {
        shouldStop = true;
        removeInt(e->downStops, e->currFloor);
    }
    
    if (shouldStop) {
        e->state = OPEN;
        e->doorTimer = OPEN_TIME;
    } else if (e->downStops->size == 0 && !hasDown(e)) {
        getNext(e);
    }
}

int unload(Elevator* e, int time) {
    int count = 0;
    PassengerNode* curr = e->passengers->head;
    PassengerNode* prev = NULL;
    
    while (curr != NULL) {
        PassengerNode* next = curr->next;
        if (curr->passenger->end == e->currFloor) {
            curr->passenger->exitTime = time;
            recordPassenger(e->sim, curr->passenger);
            
            if (prev == NULL) {
                e->passengers->head = next;
            } else {
                prev->next = next;
            }
            
            free(curr->passenger);
            free(curr);
            e->passengers->count--;
            count++;
        } else {
            prev = curr;
        }
        curr = next;
    }
    
    return count;
}

void handleDoors(Elevator* e, int time) {
    unload(e, time);
    
    Floor* f = e->sim->floors[e->currFloor];
    
    if (f != NULL && canBoard(e)) {
        bool willGoUp = e->upStops->size > 0 || hasUp(e);
        bool willGoDown = e->downStops->size > 0 || hasDown(e);
        
        while (canBoard(e) && !isQueueEmpty(f->waiting)) {
            Passenger* p = peekQueue(f->waiting);
            if (p == NULL) break;
            
            bool passengerGoingUp = p->end > p->start;
            
            if ((willGoUp && passengerGoingUp) || 
                (willGoDown && !passengerGoingUp) ||
                (!willGoUp && !willGoDown)) {
                
                int newStop = p->end;
                bool wouldCreateNewStop = false;
                
                if (passengerGoingUp) {
                    wouldCreateNewStop = !containsInt(e->upStops, newStop) && 
                                        !isTarget(e, newStop);
                } else {
                    wouldCreateNewStop = !containsInt(e->downStops, newStop) && 
                                        !isTarget(e, newStop);
                }
                
                int totalStops = e->upStops->size + e->downStops->size;
                if (wouldCreateNewStop && totalStops >= 8 && e->passengers->count > 5) {
                    break;
                }
                
                p = dequeue(f->waiting);
                p->enterTime = time;
                addPassenger(e->passengers, p);
                e->lastPickupTime = time;
                
                if (passengerGoingUp) {
                    addToArray(e->upStops, p->end);
                } else {
                    addToArray(e->downStops, p->end);
                }
                
                willGoUp = e->upStops->size > 0 || hasUp(e);
                willGoDown = e->downStops->size > 0 || hasDown(e);
            } else {
                break;
            }
        }
    }
    
    e->doorTimer--;
    if (e->doorTimer <= 0) {
        getNext(e);
    }
}

void getNext(Elevator* e) {
    bool hasUpPassengers = hasUp(e);
    bool hasDownPassengers = hasDown(e);
    
    if (hasUpPassengers || e->upStops->size > 0) {
        e->state = UP;
    } else if (hasDownPassengers || e->downStops->size > 0) {
        e->state = DOWN;
    } else {
        e->state = IDLE;
    }
}

void updateElevator(Elevator* e, int time) {
    switch (e->state) {
        case UP:
            moveUp(e, time);
            break;
        case DOWN:
            moveDown(e, time);
            break;
        case OPEN:
            handleDoors(e, time);
            break;
        case IDLE:
            updateElevState(e);
            break;
    }
}

void processRequests(ElevatorSimulation* sim, int time) {
    Floor* upRequests[NUM_FLOORS];
    Floor* downRequests[NUM_FLOORS];
    int upCount = 0;
    int downCount = 0;
    
    for (int i = 0; i < NUM_FLOORS; i++) {
        Floor* f = sim->floors[i];
        if (isQueueEmpty(f->waiting)) continue;
        
        Passenger* p = peekQueue(f->waiting);
        if (p == NULL) continue;
        
        if (p->end > p->start) {
            upRequests[upCount++] = f;
        } else {
            downRequests[downCount++] = f;
        }
    }
    
    // Simple sort by waiting count (bubble sort)
    for (int i = 0; i < upCount - 1; i++) {
        for (int j = 0; j < upCount - i - 1; j++) {
            if (upRequests[j]->waiting->count < upRequests[j+1]->waiting->count) {
                Floor* temp = upRequests[j];
                upRequests[j] = upRequests[j+1];
                upRequests[j+1] = temp;
            }
        }
    }
    
    for (int i = 0; i < downCount - 1; i++) {
        for (int j = 0; j < downCount - i - 1; j++) {
            if (downRequests[j]->waiting->count < downRequests[j+1]->waiting->count) {
                Floor* temp = downRequests[j];
                downRequests[j] = downRequests[j+1];
                downRequests[j+1] = temp;
            }
        }
    }
    
    bool assignedElevators[NUM_ELEV] = {false};
    
    // Process up requests
    for (int i = 0; i < upCount; i++) {
        Floor* f = upRequests[i];
        if (isQueueEmpty(f->waiting)) continue;
        
        Elevator* best = NULL;
        double bestScore = DBL_MAX;
        
        for (int j = 0; j < NUM_ELEV; j++) {
            if (assignedElevators[j]) continue;
            
            double score = calculateScore(sim->elevators[j], f->num, true, time);
            
            if (score < DBL_MAX && score < bestScore) {
                bestScore = score;
                best = sim->elevators[j];
            }
        }
        
        if (best != NULL && bestScore < 500) {
            addRequest(best, f->num, true);
            assignedElevators[best->num] = true;
        }
    }
    
    // Process down requests
    for (int i = 0; i < downCount; i++) {
        Floor* f = downRequests[i];
        if (isQueueEmpty(f->waiting)) continue;
        
        Elevator* best = NULL;
        double bestScore = DBL_MAX;
        
        for (int j = 0; j < NUM_ELEV; j++) {
            if (assignedElevators[j]) continue;
            
            double score = calculateScore(sim->elevators[j], f->num, false, time);
            
            if (score < DBL_MAX && score < bestScore) {
                bestScore = score;
                best = sim->elevators[j];
            }
        }
        
        if (best != NULL && bestScore < 500) {
            addRequest(best, f->num, false);
            assignedElevators[best->num] = true;
        }
    }
}

void runSimulation(ElevatorSimulation* sim) {
    srand(time(NULL));
    
    for (int t = 0; t < SIM_TIME; t++) {
        // People entering building
        int numUp = rand() % 5;
        for (int i = 0; i < numUp; i++) {
            int dest = rand() % (NUM_FLOORS - 1) + 1;
            Passenger* p = newPassenger(0, dest, t);
            enqueue(sim->floors[0]->waiting, p);
        }
        
        // People leaving building
        int numDown = rand() % 5;
        for (int i = 0; i < numDown; i++) {
            int start = rand() % (NUM_FLOORS - 1) + 1;
            Passenger* p = newPassenger(start, 0, t);
            enqueue(sim->floors[start]->waiting, p);
        }
        
        for (int i = 0; i < NUM_ELEV; i++) {
            updateElevator(sim->elevators[i], t);
        }
        
        processRequests(sim, t);
    }
}

void printResults(ElevatorSimulation* sim) {
    /*printf("Total passengers delivered: %d\n", sim->numPassengers);
    
    int totalWaiting = 0;
    for (int i = 0; i < NUM_FLOORS; i++) {
        totalWaiting += sim->floors[i]->waiting->count;
    }
    printf("Passengers still waiting: %d\n", totalWaiting);
    
    printf("\nElevator Status:\n");
    for (int i = 0; i < NUM_ELEV; i++) {
        Elevator* e = sim->elevators[i];
        const char* stateStr[] = {"IDLE", "UP", "DOWN", "OPEN"};
        printf("Elevator %d: Floor %d, State: %s, Passengers: %d, Stops: %d\n",
               e->num, e->currFloor, stateStr[e->state], 
               e->passengers->count, e->upStops->size + e->downStops->size);
    }*/
    
    if (sim->numPassengers > 0) {
        printf("\nAvg wait time: %ld sec\n", sim->totalWaitTime / sim->numPassengers);
        printf("Avg total trip time: %ld sec\n", sim->totalTravelTime / sim->numPassengers);
    }
    printf("\nAvg distance per elevator: %ld floors\n", sim->totalDistance / NUM_ELEV);
}

int main() {
    ElevatorSimulation* sim = newSim();
    
    // Initialize elevators
    for (int i = 0; i < NUM_ELEV; i++) {
        sim->elevators[i] = newElevator(i, sim);
    }
    
    runSimulation(sim);
    printResults(sim);
    
    return 0;
}